src/sage/graphs/graph_decompositions/slice_decomposition.pyx

```diff
@@ -0,0 +1,970 @@
+    vertices, `S = \{u | i \leq \sigma(u) \leq j\}`, such that for any `u \in
+    S`, we have `N_i(u) = N_i(\sigma^{-1}(i))` and for any `v` such that `j <
+    \sigma(v)`, `N_i(v) \neq N_i(\sigma^{-1}(i))`. The *head* of a part is the
+    first position of its vertices.
+
+    The algorithm starts with a single part containing all vertices. Then, when
+    the position of the `i`-th vertex `v` is fixed, it explores the neighbors of
+    `v` that have not yet been ordered. Consider a part `S` such that `N(x)\cap
+    S \neq \emptyset`. The algorithm will rearrange the ordering of the vertices
+    in `S` so that the first vertices are the neighbors of `v`. The subpart
+    containing the neighbors of `v` is assigned a new name, and the head of `S`
+    is set to the position of the first vertex of `S \setminus N(v)` in the
+    ordering `\sigma`.
+
+    Observe that each arc of the graph can induce the subdivision of a part.
+    Hence, the algorithm can use up to `m + 1` different parts.
+
+    The time complexity of this algorithm is in `O(n + m)`, and our
+    implementation follows that complexity ``SparseGraph``. For ``DenseGraph``,
+    the complexity is `O(n^2)`. See [HMPV2000]_ and [TCHP2008]_ for more
+    details.
+
+    This implementation of extended LexBFS offers some guarantee on the order in
+    which the vertices appear in the computed ordering: in case of a tie between
+    lexicographic labels during the computation, this function will "choose" the
+    vertices in the order in which they appear during the enumeration of the
+    neighbors of their last common neighbor.
+    For example, if `(u_0, ..., u_k)` is the beginning of the ordering being
+    computed and the vertices `v` and `w` currently have the same lexicographic
+    label (it means that they have the same neighbors in `(u_0, ..., u_k)`).
+    Let call `u_j` their last neighbor in the current ordering (*i.e.*, for all
+    `i > j`, `u_i` is not a neighbor of `v` and `w`). This implementation
+    will choose `v` for the next vertex of the ordering if and only if `v`
+    appeared before `w` when the neighbors of `u_j` where enumerated.
+
+    One possible use of this guarantee is that the caller can reorder the
+    adjacency list of vertices (by using, for example, a static sparse graph)
+    to force the computed LexBFS order to respect a previous one.
+
+    EXAMPLES:
+
+    To see how it can be used, see the code of the lex_BFS method (in
+    traversals.pyx) or of the class SliceDecomposition in this module.
+
+    """
+    cdef int n = <int> cg.num_verts
+    # Variables for the partition refinement algorithm
+    cdef size_t max_nparts = cg.num_arcs // 2 + 1
+    cdef bint need_to_delete_sigma_inv = sigma_inv == NULL
+    if sigma_inv == NULL:
+        sigma_inv = new vector[int]()
+    cdef vector[size_t] part_of = vector[size_t](n, 0)
+    cdef vector[size_t] part_len # only used if xslice_len != NULL (see below)
```

Comment:
2 spaces before `#`. Several places.

Reply:
done in 805d899755bf2fc2ae1c3070a6228f4d168f6f88